homeassistant/components/reolink/host.py

```diff
@@ -363,32 +364,44 @@ async def handle_webhook(
         self, hass: HomeAssistant, webhook_id: str, request: Request
     ):
         """Shield the incoming webhook callback from cancellation."""
-        await asyncio.shield(self.handle_webhook_shielded(hass, webhook_id, request))
+        self._webhook_read_done.clear()
+        asyncio.shield(self._handle_webhook(hass, webhook_id, request))
+        _LOGGER.debug("Webhook '%s' called", webhook_id)
+        await self._webhook_read_done.wait()
```

Comment:
You still need to await the future returned by `asyncio.shield`

The event likely isn't needed here because there are no `await`s after the `.set()` below

Reply:
@bdraco there is a await in the `await self._api.ONVIF_event_callback(data)` which is the handeling of the callback in the upstream library. That has other awaits in it to poll the camera if the ONVIF message is incomplete (missing AI events), has errors in it (like the source channel missing) or other issues. That is why the Event makes sense to me.

Why do we need to await the future of `asyncio.shield`?
According to https://superfastpython.com/asyncio-shield/:

> If a coroutine is provided to the asyncio.shield() function it is wrapped in an asyncio.Task() and scheduled immediately.
> This means that the shield does not need to be awaited for the inner coroutine to run.

I did test the code, and the self._handle_webhook runs just fine. Is it because of some cleanup/memory leak?

Reply:
> @bdraco there is a await in the `await self._api.ONVIF_event_callback(data)` which is the handeling of the callback in the upstream library. That has other awaits in it to poll the camera if the ONVIF message is incomplete (missing AI events), has errors in it (like the source channel missing) or other issues. That is why the Event makes sense to me.
> 

My bad. I missed that one when looking on my mobile


> Why do we need to await the future of `asyncio.shield`? According to [superfastpython.com/asyncio-shield](https://superfastpython.com/asyncio-shield/):
> 
> > If a coroutine is provided to the asyncio.shield() function it is wrapped in an asyncio.Task() and scheduled immediately.
> > This means that the shield does not need to be awaited for the inner coroutine to run.
> 
> I did test the code, and the self._handle_webhook runs just fine. Is it because of some cleanup/memory leak?

If there is an exception in the task it and you don't await it, it will never be retrieved and log a warning

Reply:
I'm still not following why you need the event though and can't await the shielded task instead

Reply:
@bdraco If the exception logging is the only problem, is it then not better to wrap the whole `_handle_webhook` in a try: except Exception (disable broad except for pylint), and in the exception Log it and re-raise it?

Then the HTTP_OK (200) can be send imediatly after reading the request.text() instead of after the possible polling which might take a second.

Reply:
There is another problem with not awaiting the task. Since nothing will hold a strong reference to it, the task could get garbage collected at any time.

https://docs.python.org/3/library/asyncio-task.html#creating-tasks

From the asyncio tasks section:
> Important Save a reference to the result of this function, to avoid a task disappearing mid-execution. The event loop only keeps weak references to tasks. A task that isn’t referenced elsewhere may get garbage collected at any time, even before it’s done. For reliable “fire-and-forget” background tasks, gather them in a collection...

Reply:
There is potentially another problem.

If you respond to the webhook before reading the contents of the message the camera could still be waiting for the buffer space to send the rest of the message which might make it drop the connection

Reply:
@bdraco the second problem is why I am using the Event and waiting on the Event untill after I have read the complete message. So that will not be a issue.

The first one about the garbage collection I was worrying about, however I do not pass in a task but am passing in the courotine directly, so asyncio.shield is creating the task internally. I have tried to figure out if garbage collection could be a issue if you do not pass a taks but a courotine directly and could not find the answer.

However I assumed the garbage collection can not be a problem if you pass in the courotine because that is a supported way of operating and if the function above would be cancelled (the reason shield exists) there would never be a refrence to the task left because you do not get such a refrence if passing in the courotine. So I am assuming asyncio.shield is taking care of this.

Do you know how this works?